client/src/main/java/com/aliyun/emr/rss/client/read/DfsPartitionReader.java

```diff
@@ -0,0 +1,178 @@
+                                location.getStorageInfo().getFilePath(), e);
+      }
+      hdfsInputStream = ShuffleClient.getHdfsFs(conf).open(
+        new Path(Utils.getSortedFilePath(location.getStorageInfo().getFilePath())));
+      getChunkOffsetsFromSortedIndex(conf, location, chunkOffsets, startMapIndex, endMapIndex);
+    } else {
+      hdfsInputStream = ShuffleClient.getHdfsFs(conf).open(
+        new Path(location.getStorageInfo().getFilePath()));
+      getChunkOffsetsFromUnsortedIndex(conf, location, chunkOffsets);
+    }
+    numChunks = chunkOffsets.size() - 1;
+    if (chunkOffsets.size() > 1) {
+      fetchThread = new Thread(() -> {
+        try {
+          int currentChunkIndex = 0;
+          while (!closed && currentChunkIndex < numChunks) {
+            while (results.size() >= maxInFlight) {
+              Thread.sleep(50);
+            }
+            long offset = chunkOffsets.get(currentChunkIndex);
+            long length = chunkOffsets.get(currentChunkIndex + 1) - offset;
+            ByteBuffer buffer = ByteBuffer.allocate((int) length);
+            hdfsInputStream.readFully(offset, buffer);
+            results.add(Unpooled.wrappedBuffer(buffer));
+            currentChunkIndex++;
+          }
+        } catch (IOException e) {
+          exception.set(e);
+        } catch (InterruptedException e) {
+          // cancel a task for speculative, ignore this exception
+        }
+      });
+      fetchThread.start();
+    }
+  }
+
+  private void getChunkOffsetsFromUnsortedIndex(RssConf conf, PartitionLocation location,
+    List<Long> chunkOffsets) throws IOException {
+    FSDataInputStream indexInputStream = ShuffleClient.getHdfsFs(conf).open(
+      new Path(Utils.getIndexFilePath(location.getStorageInfo().getFilePath())));
+    int offsetCount = indexInputStream.readInt();
+    for (int i = 0; i < offsetCount; i++) {
+      chunkOffsets.add(indexInputStream.readLong());
+    }
+    indexInputStream.close();
+  }
+
+  private void getChunkOffsetsFromSortedIndex(RssConf conf, PartitionLocation location,
+    List<Long> chunkOffsets, int startMapIndex, int endMapIndex) throws IOException {
+    String indexPath = Utils.getIndexFilePath(location.getStorageInfo().getFilePath());
+    FSDataInputStream indexInputStream = ShuffleClient.getHdfsFs(conf).open(new Path(indexPath));
+    long indexSize = ShuffleClient.getHdfsFs(conf).getFileStatus(new Path(indexPath)).getLen();
+    // Index size won't be large, so it's safe to do the conversion.
+    ByteBuffer indexBuffer = ByteBuffer.allocate((int) indexSize);
+    indexInputStream.readFully(0L, indexBuffer);
+    chunkOffsets.addAll(ShuffleBlockInfoUtils.getChunkOffsetsFromShuffleBlockInfos(startMapIndex,
+      endMapIndex, chunkSize,
+      ShuffleBlockInfoUtils.parseShuffleBlockInfosFromByteBuffer(indexBuffer)));
+    indexInputStream.close();
+  }
+
+  @Override
+  public boolean hasNext() {
+    return currentChunkIndex <= numChunks;
```

Comment:
Will currentChunkIndex and numChunks both be zero if no data exists? Then hasNext will incorrectly return true.